### PR TITLE
feat: expose assert in Deno namespace

### LINKED ISSUE #5684: Expose internal `assert` globally.
While it isn't a browser standard, Node.js includes an `assert()` globally.  We heavily use `assert()` in the internal code, but at runtime there is no `assert()` available.

Exposing it as `Deno.assert()` would be problematic in TypeScript because of https://github.com/microsoft/TypeScript/issues/37818 as it would require a type annotation all the time if you tried to destructure it and it looks like it is unlikely to ever be addressed in TypeScript because of design limitations.

We have `assert()` in `std/testing/asserts.ts` but including a testing module just to be able to do an `assert()` seems very heavy weight as it includes two other dependencies as well:

```
https://deno.land/x/std/testing/asserts.ts
  ├── https://deno.land/x/std/fmt/colors.ts
  └── https://deno.land/x/std/testing/diff.ts
```

So while it isn't ideal to have something in the global namespace that isn't in browsers, it is a minor variation that makes it significantly easier to write better code, especially since `assert()` also provides type narrowing in TypeScript now.

### PR DESCRIPTION
**Pull request by kitsonk**
Resolves #5684 

This PR exposes the internal `assert()` function globally in Deno.  It also adds `Deno.AssertionError` and `Deno.AssertionErrorOptions` and improves the internal `AssertionError` to align more to the common conventions around assertion errors in JavaScript (including an `actual`, `expected`, `operator` properties, which can be used to provide more intelligent logging of assertion errors as well as providing a mechanism to re-anchor the stack to the point where the assertion failure occurred).

This is a _breaking change_ for `std` as `std/testing/asserts.ts` uses the `Deno.AssertionError` and re-exports it.

### CODE REVIEW

**Comment by ry** [comment]
It seems `assert()` is a global in Chrome... Is there some standard that specified this? I had no idea.

**Comment by nayeemrmn** [comment]
> It seems `assert()` is a global in Chrome

I'm not seeing it... @kitsonk also said it's not in browsers. Was that in some site?

**Comment by ry** [comment]
Ah it must have just been on the site I tried. Nevermind.

**Comment by nayeemrmn** [comment]
> Exposing it as Deno.assert() would be problematic in TypeScript because of microsoft/TypeScript#37818 as it would require a type annotation all the time if you tried to destructure it and it looks like it is unlikely to ever be addressed in TypeScript because of design limitations.

If we really need a built-in `assert()`, I suggest going with `Deno.assert()` and letting this be an upstream bug. As of #6268, we don't use the destructuring convention anymore so this is definitely not worth losing compatibility over.

---

I'm against a built-in assert. Similar to #4052, it just doesn't do enough to encourage non-portable code. Why hasn't the web platform exposed an `assert()`? Can we rationalise Deno needing it more? Will an `assert` construct show up in ECMA one day?
```js
Deno.assert(predicate, "some message");
// vs
if (!predicate) throw new Error("assertion: some message");
```

**Comment by kitsonk** [comment]
I would be with you, but a couple things:

- The assert pattern is used extensively throughout the Deno internal code and `std`.  It is implied in all sorts of web specs.
- Python and C++ have it as a keyword, Rust has it as part of the very fundamental `std` crate.
- `console.assert()` logs but does not throw and is part of a specification (https://console.spec.whatwg.org/#assert)
- TypeScript supports type narrowing, so it does, in my opinion more than something like `Deno.delay()`.  It is more than syntactical sugar.
- Having `AssertionError`, which is semi-standard is also a benefit, as it makes the errors catchable.

Given that it isn't exposed globally in Node.js though, and coupled with the the focus on not destructuring, I would be comfortable moving it to `Deno.assert()` and it can be exported from `std/testing/asserts` with an explicit type annotation.

The only other solution that would be worth considering is that we have `std/assert/assert.ts` and then `std/testing/asserts.ts` which imports it, so that `assert` can be imported for functional code without dragging in half of `std`.

File: std/testing/asserts.ts [hunk 1]
```diff
@@ -133,12 +128,10 @@ export function equal(c: unknown, d: unknown): boolean {
...
-/** Make an assertion, if not `true`, then throw. */
-export function assert(expr: unknown, msg = ""): asserts expr {
-  if (!expr) {
-    throw new AssertionError(msg);
-  }
-}
+export const assert: (
+  condition: unknown,
+  message?: string | undefined
+) => asserts condition = Deno.assert;
```
Comment (SyrupThinker) [inline std/testing/asserts.ts:134]: Unconditionally assigning `Deno.assert` makes the asserts module incompatible with other runtimes.
Maybe it should be done like in line 18 to preserve compatibility?

**Comment by bartlomieju** [comment]
I think we're in agreement that we don't want to add `assert()` global that is not present in browser as it would break the compatibility promise we made. I guess we just have to keep status quo for now? @ry @kitsonk @nayeemrmn

**Comment by kitsonk** [comment]
> I guess we just have to keep status quo for now?

While this PR would need to be rewritten, I am still strongly in favour of exposing the internal `assert()` and `AssertionError` as `Deno.assert()` and `Deno.AssertionError` (which is what this PR did before change in the cli's runtime code).  I am still really tired of dragging in a half dozen modules from `std` to be able to use `assert()` in Deno.

**Comment by ry** [comment]
I'm not keen to expand the Deno namespace with pure JS functionality, because then any code that uses `Deno.assert` instead of `if (!x) throw Error("fail")` is unnecessarily made browser incompatible. With all the functionality the browser provides, I also wish it provided a proper assert function. It would be so easy. But it doesn't.

A similar case could be made about the `createResolvable()` function that has been defined everywhere. Where does it stop?

I understand the desire, but in the interest of moving Deno towards browser compatibility, I don't think we should land this.

**Comment by kitsonk** [comment]
Ok, I will focus on trying to improve `std/testing` and asserts then, because again, we shouldn't require a load of modules just to have an assert function that throws an assertion error.

**Pull request closed without merging**